### PR TITLE
Add TestGroup to siatest package

### PR DESCRIPTION
**Pull request by ChrisSchinnerl**
This PR adds the TestGroup struct to the new siatest package. It allows users to spin up large groups of nodes ready for tests with minimal effort. 
This will allow us to easily create more reliable tests for the upcoming downloader overhaul and various other PRs that are currently queued for testing.

### CODE REVIEW

File: siatest/testgroup.go [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+		miners  map[*TestNode]struct{}
+
+		dir string
+	}
+)
+
+var (
+	// defaultAllowance is the allowance used for the group's renters
+	defaultAllowance = modules.Allowance{
+		Funds:       types.SiacoinPrecision.Mul64(100000),
```
Comment (DavidVorick) [inline siatest/testgroup.go]: 100e3

File: siatest/testgroup.go [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+		Hosts:       10,
```
Comment (DavidVorick) [inline siatest/testgroup.go]: 4

Comment (DavidVorick) [reply siatest/testgroup.go]: Eh actually 5

File: siatest/testgroup.go [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+		RenewWindow: 5,
+	}
+)
+
+// announceHosts adds storage to each host and announces them to the group
+func announceHosts(hosts map[*TestNode]struct{}) error {
+	for host := range hosts {
+		if err := host.HostStorageFoldersAddPost(host.Dir, 1048576); err != nil {
+			return err
+		}
```
Comment (DavidVorick) [inline siatest/testgroup.go]: Adding a storage folder should be separate from announcing

File: siatest/testgroup.go [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+	return
+}
+
+// fullyConnectNodes takes a list of nodes and connects all their gateways
+func fullyConnectNodes(nodes []*TestNode) error {
+	// Fully connect the nodes
+	for i, nodeA := range nodes {
+		for _, nodeB := range nodes[i+1:] {
+			if err := nodeA.GatewayConnectPost(nodeB.GatewayAddress()); err != nil {
+				return build.ExtendErr("failed to connect to peer", err)
```
Comment (DavidVorick) [inline siatest/testgroup.go]: Is it possible that this returns 'already a peer' errors?

File: siatest/testgroup.go [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+
+// waitForContracts waits until the renters have formed contracts with the
+// hosts in the group.
+func waitForContracts(miner *TestNode, renters map[*TestNode]struct{}, hosts map[*TestNode]struct{}) error {
+	expectedContracts := defaultAllowance.Hosts
+	if uint64(len(hosts)) < expectedContracts {
+		expectedContracts = uint64(len(hosts))
+	}
+	for renter := range renters {
+		err := build.Retry(100, time.Second, func() error {
```
Comment (DavidVorick) [inline siatest/testgroup.go]: No need to wait a full second between tries, can wait 100ms

File: siatest/testgroup_test.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+	}
+}
+
+// TestCreateTestGroup tests NewGroup with no renter or host
+func TestNewGroupNoRenterHost(t *testing.T) {
+	if testing.Short() {
+		t.SkipNow()
+	}
+	// Create a group with nothing but a single miner
+	groupParams := GroupParams{
```
Comment (DavidVorick) [inline siatest/testgroup_test.go]: "a single miner"

miners: 5

**Comment by DavidVorick** [comment]
Overall this looks pretty good

File: siatest/testgroup.go (hunk 4 above)
Comment (ChrisSchinnerl) [reply siatest/testgroup.go]: Hm... that could probably happen if you have the nodes A, B and C and they connect in the following order: AB, AC, BC. 
B and C might already have connected due to being connect via A. This is probably not very likely for small groups but I will add a check.

File: siatest/testgroup.go (hunk 5 above)
Comment (ChrisSchinnerl) [reply siatest/testgroup.go]: I would recommend to leave it at 1 second and add a todo. Forming contracts is incredibly slow at the moment. If we change it to 100 milliseconds, 100 tries are not enough. That is not a bad thing usually but we might not want to mine >100 blocks while waiting for the renters to form contracts.

File: siatest/testgroup.go [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+		}
+		if err := host.HostAnnouncePost(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Close closes the group and all its nodes
+func (tg *TestGroup) Close() (err error) {
```
Comment (DavidVorick) [inline siatest/testgroup.go]: All methods of a type should be grouped together.

Beyond that, we try to keep everything alphabetically sorted.

File: siatest/testgroup.go [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+		if err := renter.RenterPost(defaultAllowance); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// NewGroupFromTemplate will create hosts, renters and miners according to the
+// settings in groupParams.
+func NewGroupFromTemplate(groupParams GroupParams) (*TestGroup, error) {
```
Comment (DavidVorick) [inline siatest/testgroup.go]: We try to cluster the exported functions (even that don't act on an object) together the same way that we try to cluster functions of an object together.

And by extension, we cluster exported functions of an object together, and the unexported functions of an object are clustered together in a separate cluster.

File: siatest/testgroup_test.go [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+	// Create the group
+	_, err := NewGroupFromTemplate(groupParams)
+	if err == nil {
+		t.Fatal("Creating a group without miners should fail: ", err)
+	}
+}
+
+// TestCreateTestGroup tests NewGroup with no renter or host
+func TestNewGroupNoRenterHost(t *testing.T) {
+	if testing.Short() {
```
Comment (DavidVorick) [inline siatest/testgroup_test.go]: This can probably be vlong

File: siatest/testgroup_test.go [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+		}
+		if len(wtg.ConfirmedTransactions) == 0 {
+			t.Errorf("Node has 0 confirmed funds")
+		}
+	}
+}
+
+// TestCreateTestGroup tests NewGroup without a miner
+func TestNewGroupNoMiner(t *testing.T) {
+	if testing.Short() {
```
Comment (DavidVorick) [inline siatest/testgroup_test.go]: also vlong

**Pull request merged**